drivers/counter/rtc_mcp7940n.c

```diff
@@ -0,0 +1,785 @@
+	if (rc < 0) {
+		goto out;
+	}
+
+	/* Config user data and callback */
+	data->counter_handler[id] = alarm_cfg->callback;
+	data->counter_ticks[id] = current_time;
+	data->alarm_user_data[id] = alarm_cfg->user_data;
+
+out:
+	k_sem_give(&data->lock);
+
+	return rc;
+}
+
+static int mcp7940n_counter_cancel_alarm(const struct device *dev, uint8_t alarm_id)
+{
+	struct mcp7940n_data *data = dev->data;
+	int rc = 0;
+
+	k_sem_take(&data->lock, K_FOREVER);
+
+	/* Clear alarm enable bit */
+	if (alarm_id == ALARM0_ID) {
+		data->registers.rtc_control.alm0_en = 0;
+	} else if (alarm_id == ALARM1_ID) {
+		data->registers.rtc_control.alm1_en = 0;
+	} else {
+		rc = -EINVAL;
+		goto out;
+	}
+
+	rc = write_register(dev, REG_RTC_CONTROL, data->registers.rtc_control);
+
+out:
+	k_sem_give(&data->lock);
+
+	return rc;
+}
+
+static int mcp7940n_counter_set_top_value(const struct device *dev,
+					  const struct counter_top_cfg *cfg)
+{
+	return -ENOTSUP;
+}
+
+/* This function can be used to poll the alarm interrupt flags if the MCU is
+ * not connected to the MC7940N MFP pin. It can also be used to check if an
+ * alarm was triggered while the MCU was in reset. This function will clear
+ * the interrupt flag
+ *
+ * Return bitmask of alarm interrupt flag (IF) where each IF is shifted by
+ * the alarm ID.
+ */
+static int mcp7940n_counter_get_pending_int(const struct device *dev)
+{
+	struct mcp7940n_data *data = dev->data;
+	uint8_t interrupt_pending = 0;
+	int rc;
+
+	k_sem_take(&data->lock, K_FOREVER);
+
+	/* Check interrupt flag for alarm 0 */
+	rc = read_register(dev, REG_ALM0_WDAY,
+		(uint8_t *)&data->alm0_registers.alm_weekday);
+	if (rc < 0) {
+		goto out;
+	}
+
+	if (data->alm0_registers.alm_weekday.alm_if) {
+		/* Clear interrupt */
+		data->alm0_registers.alm_weekday.alm_if = 0;
+		rc = write_register(dev, REG_ALM0_WDAY,
+			data->alm0_registers.alm_weekday);
+		if (rc < 0) {
+			goto out;
+		}
+		interrupt_pending |= (1 << ALARM0_ID);
+	}
+
+	/* Check interrupt flag for alarm 1 */
+	rc = read_register(dev, REG_ALM1_WDAY,
+		(uint8_t *)&data->alm1_registers.alm_weekday);
+	if (rc < 0) {
+		goto out;
+	}
+
+	if (data->alm1_registers.alm_weekday.alm_if) {
+		/* Clear interrupt */
+		data->alm1_registers.alm_weekday.alm_if = 0;
+		rc = write_register(dev, REG_ALM1_WDAY,
+			data->alm1_registers.alm_weekday);
+		if (rc < 0) {
+			goto out;
+		}
+		interrupt_pending |= (1 << ALARM1_ID);
+	}
+
+out:
+	k_sem_give(&data->lock);
+
+	if (rc == 0) {
+		rc = (int)interrupt_pending;
+	}
+	return rc;
+}
+
+static int mcp7940n_counter_get_top_value(const struct device *dev)
+{
+	return UINT32_MAX;
+}
+
+static int mcp7940n_init(const struct device *dev)
+{
+	struct mcp7940n_data *data = dev->data;
+	const struct mcp7940n_config *cfg = dev->config;
+	int rc = 0;
+	time_t unix_time = 0;
+
+	/* Initialize and take the lock */
+	k_sem_init(&data->lock, 0, 1);
+
+	if (!device_is_ready(cfg->i2c_dev)) {
+		LOG_ERR("I2C device not found");
```

Comment:
"not ready" is not the same as "not found". Please use `LOG_ERR("I2C device %s is not ready", cfg->i2c_dev->name);`

Reply:
Updated.